src/utils/hashing.ts

```diff
@@ -2,229 +2,295 @@
-// representing a number of bit positions, returns the 64bit int (as an
-// array of two 32bit ints) shifted left by that number of positions.
-//
-function x64LeftShift(m: number[], n: number) {
-  n %= 64
-  if (n === 0) {
-    return m
-  } else if (n < 32) {
-    return [(m[0] << n) | (m[1] >>> (32 - n)), m[1] << n]
+/**
+ * Provides a left shift of the given int32 value (provided as tuple of [0, int32])
+ * by given number of bits. Result is written back to the value
+ */
+function x64LeftShift(m: number[], bits: number): void {
+  bits %= 64
+  if (bits === 0) {
+    return
+  } else if (bits < 32) {
+    m[0] = m[1] >>> (32 - bits)
+    m[1] = m[1] << bits
   } else {
-    return [m[1] << (n - 32), 0]
+    m[0] = m[1] << (bits - 32)
+    m[1] = 0
   }
 }
-//
-// Given two 64bit ints (as an array of two 32bit ints) returns the two
-// xored together as a 64bit int (as an array of two 32bit ints).
-//
-function x64Xor(m: number[], n: number[]) {
-  return [m[0] ^ n[0], m[1] ^ n[1]]
+/**
+ * Provides a XOR of the given int64 values(provided as tuple of two int32).
+ * Result is written back to the first value
+ */
+function x64Xor(m: number[], n: number[]): void {
+  m[0] ^= n[0]
+  m[1] ^= n[1]
 }
-//
-// Given a block, returns murmurHash3's final x64 mix of that block.
-// (`[0, h[0] >>> 1]` is a 33 bit unsigned right shift. This is the
-// only place where we need to right shift 64bit ints.)
-//
-function x64Fmix(h: number[]) {
-  h = x64Xor(h, [0, h[0] >>> 1])
-  h = x64Multiply(h, [0xff51afd7, 0xed558ccd])
-  h = x64Xor(h, [0, h[0] >>> 1])
-  h = x64Multiply(h, [0xc4ceb9fe, 0x1a85ec53])
-  h = x64Xor(h, [0, h[0] >>> 1])
-  return h
+
+const F1 = [0xff51afd7, 0xed558ccd]
+const F2 = [0xc4ceb9fe, 0x1a85ec53]
+/**
+ * Calculates murmurHash3's final x64 mix of that block and writes result back to the input value.
+ * (`[0, h[0] >>> 1]` is a 33 bit unsigned right shift. This is the
+ * only place where we need to right shift 64bit ints.)
+ */
+function x64Fmix(h: number[]): void {
+  const shifted = [0, h[0] >>> 1]
+  x64Xor(h, shifted)
+  x64Multiply(h, F1)
+  shifted[1] = h[0] >>> 1
+  x64Xor(h, shifted)
+  x64Multiply(h, F2)
+  shifted[1] = h[0] >>> 1
+  x64Xor(h, shifted)
 }
 
-//
-// Given a string and an optional seed as an int, returns a 128 bit
-// hash using the x64 flavor of MurmurHash3, as an unsigned hex.
-//
-export function x64hash128(key: string, seed?: number): string {
-  key = key || ''
+const C1 = [0x87c37b91, 0x114253d5]
+const C2 = [0x4cf5ad43, 0x2745937f]
+const M = [0, 5]
+const N1 = [0, 0x52dce729]
+const N2 = [0, 0x38495ab5]
+/**
+ * Given a string and an optional seed as an int, returns a 128 bit
+ * hash using the x64 flavor of MurmurHash3, as an unsigned hex.
+ * All internal functions mutates passed value to achieve minimal memory allocations and GC load
+ *
+ * Benchmark https://jsbench.me/p4lkpaoabi/1
+ */
+export function x64hash128(input: string, seed?: number): string {
+  const key = getUTF8Bytes(input)
```

Comment:
Just an idea, probably will give a minor or no effect: iterate over the blocks of 16 bytes instead of making a full array of bytes.

Reply:
Do you mean to collect bytes while iterating through unicode codepoints until we have 16-bytes block?

Reply:
Yes, because as I can see, the algorithm operates 16-byte blocks

Reply:
For history: this was a good exersice but the approach is slower (maybe my custom implementation just can't beat some `TextEncoder` optimizations)

Code points implementation https://github.com/fingerprintjs/fingerprintjs/pull/936